compiler/shared_ast/typing.ml

```diff
@@ -312,13 +312,9 @@ let polymorphic_op_type (op : Operator.polymorphic A.operator Mark.pos) :
     | Log (PosRecordIfTrueBool, _) -> [bt] @-> bt
     | Log _ -> [any] @-> any
     | Length -> [array any] @-> it
-    (* The [HandleDefault] and [HandleDefaultOpt] need to be typed before and
-       after the Lcalc monomorphization which affects arrays and option types.
-       Because of that, we give the operators very lax typing rules with [any]
-       but it doesn't matter for unification because the concrete types on which
-       they will be instantiated are stored in the [EAppOp] node. *)
-    | HandleDefault -> [any2; [ut] @-> bt; [ut] @-> any] @-> any
-    | HandleDefaultOpt -> [any2; [ut] @-> bt; [ut] @-> any] @-> any
+    | HandleDefault ->
```

Comment:
This will break the C backend for sure, because the monomorphization pass changes the `Array` type and expression (to make monomorphic array). In this case, the type of `HandleDefaultOpt` is no longer `array any -> ...` but `(struct monomorphized_array) -> ...`. This is why I put `any2` in here.

Reply:
Hm that's right... what I don't understand though is how the other operators working on arrays can be typed then ? Or have they disappeared at this point ?

Another interrogation is wether we really need to re-infer types after monomorphisation. If yes, it might be worth wondering if we could switch to a different typing for operators at that point.

An sneaky option could, maybe, to skip that function if the full annotations are already in place.

Of course, I am exploring things here but don't want to hinder progress on the C branch: feel free to ignore any part of this to go on, and this kind of improvements can be kept for later if needed at all.

Reply:
The other operators working on arrays will also have to have their typing tweaked here, it's just that I did not reach the point in the `c_backend` PR where I'm dealing with them.

Reply:
Yes we need to re-infer types after monomorphization. The reason for that is that it helps me debug my monomorphization pass. Furthermore, right now all my monomorphization tweaks are suhch that I can manually put the new type in the resulting terms, but that may not be true for future monomorphization tweaks.